ykllvmwrap/src/jitbuilder.cc

```diff
@@ -0,0 +1,315 @@
+    llvm::FunctionType *FType =
+        llvm::FunctionType::get(Type::getVoidTy(JITContext), InputTypes, false);
+    llvm::Function *JITFunc = llvm::Function::Create(
+        FType, Function::InternalLinkage, TraceName, JITMod);
+    JITFunc->setCallingConv(CallingConv::C);
+
+    // Create entry block and setup builder.
+    auto DstBB = BasicBlock::Create(JITContext, "", JITFunc);
+    llvm::IRBuilder<> Builder(JITContext);
+    Builder.SetInsertPoint(DstBB);
+
+    // Variables that are used (but not defined) inbetween start and stop
+    // tracing need to be replaced with function arguments which the user passes
+    // into the compiled trace. This loop creates a mapping from those original
+    // variables to the function arguments of the compiled trace function.
+    for (size_t Idx = 0; Idx != Inputs.size(); Idx++) {
+      Value *OldVal = Inputs[Idx];
+      Value *NewVal = JITFunc->getArg(Idx);
+      assert(NewVal->getType()->isPointerTy());
+      VMap[OldVal] = NewVal;
+    }
+
+    std::vector<CallInst *> inlined_calls;
+    CallInst *last_call = nullptr;
+
+    // Iterate over the PT trace and stitch together all traced blocks.
+    for (size_t Idx = 0; Idx < Len; Idx++) {
+      auto FuncName = FuncNames[Idx];
+
+      // FIXME: Deal with holes in the trace, e.g. calls to libc.
+      assert(FuncName != nullptr);
+
+      // Get a traced function so we can extract blocks from it.
+      Function *F = AOTMod->getFunction(FuncName);
+      if (!F)
+        errx(EXIT_FAILURE, "can't find function %s", FuncName);
+
+      // Skip to the correct block.
+      auto It = F->begin();
+      std::advance(It, BBs[Idx]);
+      BasicBlock *BB = &*It;
+
+      // Iterate over all instructions within this block and copy them over
+      // to our new module.
+      for (auto I = BB->begin(); I != BB->end(); I++) {
+        // Skip calls to debug intrinsics (e.g. @llvm.dbg.value). We don't
+        // currently handle debug info and these "pseudo-calls" cause our blocks
+        // to be prematurely terminated.
+        if (isa<DbgInfoIntrinsic>(I))
+          continue;
+
+        // If we've returned from a call, skip ahead to the instruction where
+        // we left off.
+        if (last_call != nullptr) {
+          if (&*I == last_call) {
+            last_call = nullptr;
+          }
+          continue;
+        }
+        if (isa<CallInst>(I)) {
+          CallInst *CI = cast<CallInst>(&*I);
+          Function *CF = CI->getCalledFunction();
+          if (CF == nullptr)
+            continue;
+
+          if (CF->getName() == YKTRACE_START) {
+            StartTracingInstr = &*I;
+            continue;
+          } else if (CF->getName() == YKTRACE_STOP) {
+            finalise(&Builder);
+            return JITMod;
+          } else {
+            // Skip remainder of this block and remember where we stopped so we
+            // can continue tracing from this position after returning frome the
```

Comment:
s/tracing// ?

Reply:
frome

Reply:
@vext01 is correct. Proper nouns should be capitalised. https://en.wikipedia.org/wiki/Frome

Reply:
Fixed in cdc0a33.